plugin/kubernetes/xfr.go

```diff
@@ -1,17 +1,81 @@
 package kubernetes
 
 import (
-	"time"
+	"strings"
 
+	"github.com/coredns/coredns/plugin/etcd/msg"
 	"github.com/coredns/coredns/request"
+
+	"github.com/miekg/dns"
+	api "k8s.io/client-go/pkg/api/v1"
 )
 
 // Serial implements the Transferer interface.
-func (e *Kubernetes) Serial(state request.Request) uint32 {
-	return uint32(time.Now().Unix())
-}
+func (k *Kubernetes) Serial(state request.Request) uint32 { return uint32(k.APIConn.Modified()) }
 
 // MinTTL implements the Transferer interface.
-func (e *Kubernetes) MinTTL(state request.Request) uint32 {
-	return 30
+func (k *Kubernetes) MinTTL(state request.Request) uint32 { return 30 }
+
+// Transfer implements the Transferer interface.
+func (k *Kubernetes) Transfer(state request.Request) <-chan msg.Service {
+	c := make(chan msg.Service)
+
+	go k.transfer(c, state.Zone)
+
+	return c
+}
+
+func (k *Kubernetes) transfer(c chan msg.Service, zone string) {
+
+	defer close(c)
+
+	zonePath := msg.Path(zone, "coredns")
+	serviceList := k.APIConn.ServiceList()
+	endpointsList := k.APIConn.EndpointsList()
+	for _, svc := range serviceList {
+		// Endpoint query or headless service
+		if svc.Spec.ClusterIP == api.ClusterIPNone {
+			for _, ep := range endpointsList {
```

Comment:
Should use the index here instead of scanning whole endpoint list for each service...  e.g. `endpointsIndex(svc.Name+"."+ svc.Namespace)`